contracts/cw-distributor/src/contract.rs

```diff
@@ -0,0 +1,130 @@
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let owner = deps.api.addr_validate(&msg.owner)?;
+    let recipient = deps.api.addr_validate(&msg.recipient)?;
+
+    let reward_token = deps.api.addr_validate(&msg.reward_token)?;
+    let config = Config {
+        owner,
+        recipient,
+        reward_token,
+        reward_rate: msg.reward_rate,
+    };
+    CONFIG.save(deps.storage, &config)?;
+
+    // Initialize last payment block
+    LAST_PAYMENT_BLOCK.save(deps.storage, &env.block.height)?;
+
+    Ok(Response::new().add_attribute("action", "instantiate"))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig {
+            owner,
+            recipient,
+            reward_rate,
+            reward_token,
+        } => execute_update_config(deps, info, owner, recipient, reward_rate, reward_token),
+        ExecuteMsg::Distribute {} => execute_distribute(deps, env),
+    }
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    info: MessageInfo,
+    owner: String,
+    recipient: String,
+    reward_rate: Uint128,
+    reward_token: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.owner != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let owner = deps.api.addr_validate(&owner)?;
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let reward_token = deps.api.addr_validate(&reward_token)?;
+
+    let config = Config {
+        owner,
+        recipient,
+        reward_token,
+        reward_rate,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    Ok(Response::default())
+}
+
+pub fn execute_distribute(deps: DepsMut, env: Env) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let last_payment_block = LAST_PAYMENT_BLOCK.load(deps.storage)?;
+    let block_diff = env.block.height - last_payment_block;
+    let amount: Uint128 = config.reward_rate * Uint128::new(block_diff.into());
+
+    if amount == Uint128::zero() {
+        return Err(ContractError::NoPendingPayments {});
+    }
+
+    LAST_PAYMENT_BLOCK.save(deps.storage, &env.block.height)?;
+
+    let msg = to_binary(&cw20::Cw20ExecuteMsg::Send {
+        contract: config.recipient.clone().into_string(),
+        amount,
+        msg: to_binary(&ReceiveMsg::Fund {}).unwrap(),
+    })?;
+    let send_msg: CosmosMsg = WasmMsg::Execute {
+        contract_addr: config.reward_token.into(),
+        msg,
+        funds: vec![],
+    }
+    .into();
+    Ok(Response::default().add_message(send_msg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
```

Comment:
I would consider combining all of this into one `info` response and including the contracts current balance of the reward token. Having reward token balance makes it easy for the caller to see if the rewards are active and calculate the remaining length without making a seperate query.

Reply:
Done